File: src/main/java/june_2025/GraphValidTree.java
package june_2025;

import java.util.*;

public class GraphValidTree {

    public static void main(String[] args) {
        int n = 5;
        int[][] edges = {{0,1}, {0,2}, {0,3}, {1,4}};

        GraphValidTree graphValidTree = new GraphValidTree();
        boolean result = graphValidTree.validTree(n, edges);
        System.out.println(result);
    }

    // O(V + E) time | O(V + E) space
    public boolean validTree(int n, int[][] edges) {
        Map<Integer, List<Integer>> adj = new HashMap<>();
        for (int i = 0; i < n; i++) {
            adj.put(i, new ArrayList<>());
        }
        for (int[] edge : edges) {
            int s = edge[0];
            int d = edge[1];
            adj.get(s).add(d);
            adj.get(d).add(s);
        }
        Set<Integer> visited = new HashSet<>();
        if (cycle(adj, 0, -1, visited)) {
            return false;
        }
        return visited.size() == n;
    }

    private boolean cycle(Map<Integer, List<Integer>> adj, int node, int prev, Set<Integer> visited) {
        if (visited.contains(node)) {
            return false;
        }

        List<Integer> neighbors = adj.get(node);
        for (int nei : neighbors) {
            if (nei == prev) {
                continue;
            }
            if (cycle(adj, nei, node, visited)) {
                return true;
            }
        }

        visited.add(node);
        return false;
    }

}

File: python/graph_valid_tree.py
class Solution:
    # O(V + E) time | O(V + E) space
    def validTree(self, n, edges):
        adj = {}
        for i in range(n):
            adj[i] = []
        
        for edge in edges:
            s = edge[0]
            d = edge[1]
            adj[s].append(d)
            adj[d].append(s)
        
        visited = set()

        def cycle(node, prev):
            if node in visited:
                return True
            
            visited.add(node)
            for nei in adj[node]:
                if nei == prev:
                    continue
                if cycle(nei, node):
                    return True
            
            return False

        if (cycle(0, -1)):
            return False

        return len(visited) == n

# n = 5
# edges = [[0, 1], [0, 2], [0, 3], [1, 4]]

n = 5
edges = [[0, 1], [1, 2], [2, 3], [1, 3], [1, 4]]

solution = Solution()
print(solution.validTree(n, edges))
        